kbase-extension/static/kbase/js/util/icon.js

```diff
@@ -1,150 +1,246 @@
+    }
+
+    /**
+     * @method _makeStackLayers
+     * @private
+     * Amends an exist array of icon layers to add in stack layers
+     * @param {string} shape      shape for the icon background;
+     *                              typically 'square' or 'circle'
+     * @param {string} color      colour for the icon background (hex or RGB)
+     *
+     * @returns {array}           array of HTML strings representing the extra icon layers
+     */
+
+    function _makeStackLayers(shape = 'square', color = null) {
+        // For 'stacked' (set) icons, add a shifted-over
+        // circle first, as the bottom layer, then also add a border
+        // to the top one.
+        const layers = [],
+            cmax = function(x) { return x > 255 ? 255 : x; },
+            cstep = 20, // color-step for overlapped circles
+            num_stacked_circles = 1; // up to 2
+        let parsed_color, r, g, b;
+
+        // XXX: Assume color is in form '#RRGGBB'
+        if (color[0] == '#') {
+            parsed_color = color.match(/#(..)(..)(..)/);
+            r = parseInt(parsed_color[1], 16);
+            g = parseInt(parsed_color[2], 16);
+            b = parseInt(parsed_color[3], 16);
+        }
+        // XXX: Assume color is in form "rgb(#,#,#)"
+        else {
+            parsed_color = color.match(/rgb\s*\(\s*(\d+)\s*,\s*(\d+)\s*,\s*(\d+)\s*\)/);
+            r = parsed_color[1];
+            g = parsed_color[2];
+            b = parsed_color[3];
+        }
+        // Add circles with lighter colors
+        for (let i=num_stacked_circles; i > 0; i--) {
+            let stacked_color = color
+                ? 'rgb(' + cmax(r + i * cstep)  + ',' +
+                    cmax(g + i * cstep) + ',' + cmax(b + i * cstep) + ')'
+                : '';
+            let style = color ? { color: stacked_color } : {};
+            layers.push(
+                span({
+                    class: `${cssBaseName}__stack--l${i} fa fa-stack-2x fa-${shape}`,
+                    style: style,
+                }),
+                // white separator
+                span({
+                    class: `${cssBaseName}__outline--l${i} fa fa-stack-2x fa-${shape}`,
+                })
+            );
+        }
+        return layers;
+    }
+
+    /**
+     * @method _makeIcon
+     * @private
+     * Creates an HTML string that renders an icon
+     * @param {string} iconType  the type of icon to produce,
+     *                              e.g. 'app', 'app-toolbar', 'data', 'generic', etc.
+     * @param {object} iconClass  the class of the icon, e.g. 'fa-cube'
+     * @param {string} shape      shape for the icon background;
+     *                              typically 'square' or 'circle'
+     * @param {string} color      colour for the icon background (hex or RGB)
+     * @param {bool}   stacked    whether or not the icon should be stacked
+     */
+
+    function _makeIcon(iconType, iconClass, shape = 'square', color = null, stacked = false) {
```

Comment:
`_makeIcon`, `_makeIconFromUrl`, and `_makeStackLayers` do all the work in this module. The other functions provide ways to call these functions with different inputs.